cloudinit/net/tests/test_networkd.py

```diff
@@ -0,0 +1,82 @@
+# This file is part of cloud-init. See LICENSE file for license information.
+
+from cloudinit import safeyaml
+from cloudinit.net import networkd, network_state
+
+V2_CONFIG_SET_NAME = """\
+network:
+  version: 2
+  ethernets:
+    eth0:
+      match:
+        macaddress: '00:11:22:33:44:55'
+      nameservers:
+        search: [spam.local, eggs.local]
+        addresses: [8.8.8.8]
+    eth1:
+      match:
+        macaddress: '66:77:88:99:00:11'
+      set-name: "ens92"
+      nameservers:
+        search: [foo.local, bar.local]
+        addresses: [4.4.4.4]
+"""
+
+V2_CONFIG_SET_NAME_RENDERED_ETH0 = """[Match]
+MACAddress=00:11:22:33:44:55
+Name=eth0
+
+[Network]
+DHCP=no
+DNS=8.8.8.8
+Domains=spam.local eggs.local
+
+"""
+
+V2_CONFIG_SET_NAME_RENDERED_ETH1 = """[Match]
+MACAddress=66:77:88:99:00:11
+Name=ens92
+
+[Network]
+DHCP=no
+DNS=4.4.4.4
+Domains=foo.local bar.local
+
+"""
+
+
+class TestNetworkdRenderState:
+    def _parse_network_state_from_config(self, config):
+        yaml = safeyaml.load(config)
+        return network_state.parse_net_config_data(yaml["network"])
+
+    def _normalize(self, str):
+        return "".join(sorted(str))
+
```

Comment:
I don't think you are testing what you think you are here:
_normalize of V2_CONFIG_SET_NAME_RENDERED_ETH1  is giving you "\n\n\n\n\n\n\n\n\n .....001124444667788999:::::=====AACCDDDHMMNNNPS[[]]aaaaaabcccddeeeefhikllllmmnnnooooooorrrssssttw..."

So essentially you are comparing whether the full string has equivalent character counts, not that it actually has the right content.

Cn we instead make sure we are sorting the content we are rendering so different versions of python we assert proper sorted order always for each section?
Here's a suggestion for your consideration:

```diff
diff --git a/cloudinit/net/networkd.py b/cloudinit/net/networkd.py
index 3b7a0edc..c97c18f6 100644
--- a/cloudinit/net/networkd.py
+++ b/cloudinit/net/networkd.py
@@ -41,11 +41,11 @@ class CfgParser:
 
     def get_final_conf(self):
         contents = ''
-        for k, v in self.conf_dict.items():
+        for k, v in sorted(self.conf_dict.items()):
             if not v:
                 continue
             contents += '['+k+']\n'
-            for e in v:
+            for e in sorted(v):
                 contents += e + '\n'
             contents += '\n'
 
diff --git a/cloudinit/net/tests/test_networkd.py b/cloudinit/net/tests/test_networkd.py
index e8f6a896..4c03e281 100644
--- a/cloudinit/net/tests/test_networkd.py
+++ b/cloudinit/net/tests/test_networkd.py
@@ -50,9 +50,6 @@ class TestNetworkdRenderState:
         yaml = safeyaml.load(config)
         return network_state.parse_net_config_data(yaml["network"])
 
-    def _normalize(self, str):
-        return "".join(sorted(str))
-
     def test_networkd_render_with_set_name(self):
         """
         Please note the _normalize function is used when comparing the
@@ -70,13 +67,9 @@ class TestNetworkdRenderState:
         rendered_content = renderer._render_content(ns)
 
         assert "eth0" in rendered_content
-        assert self._normalize(rendered_content["eth0"]) == self._normalize(
-            V2_CONFIG_SET_NAME_RENDERED_ETH0
-        )
+        assert rendered_content["eth0"] == V2_CONFIG_SET_NAME_RENDERED_ETH0
         assert "ens92" in rendered_content
-        assert self._normalize(rendered_content["ens92"]) == self._normalize(
-            V2_CONFIG_SET_NAME_RENDERED_ETH1
-        )
+        assert rendered_content["ens92"] == V2_CONFIG_SET_NAME_RENDERED_ETH1
 
 
 # vi: ts=4 expandtab
```

Don't we instead want to

Reply:
Ack.